### PR TITLE
Simulator Development

### PR DESCRIPTION
**Pull request by janeusz2000**
this is still in development. I made this for you to view changes that are happing. There is a lot to be done and Im not happy with the design yet

### CODE REVIEW

**Comment by janeusz2000** [comment]
image
my tests are all passed so this is weid. I will start fixing tommorow ;)

**Comment by dejw** [comment]
> image
> my tests are all passed so this is weid. I will start fixing tommorow ;)


Have you read the error message in the build output?

It is this:

> bazel-out/k8-opt/bin/_solib_k8/liblibutilsSimulator.so: error: undefined reference to 'SceneManager::getSphereWall()'
> bazel-out/k8-opt/bin/_solib_k8/liblibutilsSimulator.so: error: undefined reference to 'SceneManager::getReferencePlate() const'

So you do not have impl. done yet.

I think I told you before, but VS is more permissive than gcc so you see those errors and warnings (warnings are especially useful).

**Comment by janeusz2000** [comment]
> > image
> > my tests are all passed so this is weid. I will start fixing tommorow ;)
> 
> Have you read the error message in the build output?
> 
> It is this:
> 
> > bazel-out/k8-opt/bin/_solib_k8/liblibutilsSimulator.so: error: undefined reference to 'SceneManager::getSphereWall()'
> > bazel-out/k8-opt/bin/_solib_k8/liblibutilsSimulator.so: error: undefined reference to 'SceneManager::getReferencePlate() const'
> 
> So you do not have impl. done yet.
> 
> I think I told you before, but VS is more permissive than gcc so you see those errors and warnings (warnings are especially useful).



> > image
> > my tests are all passed so this is weid. I will start fixing tommorow ;)
> 
> Have you read the error message in the build output?
> 
> It is this:
> 
> > bazel-out/k8-opt/bin/_solib_k8/liblibutilsSimulator.so: error: undefined reference to 'SceneManager::getSphereWall()'
> > bazel-out/k8-opt/bin/_solib_k8/liblibutilsSimulator.so: error: undefined reference to 'SceneManager::getReferencePlate() const'
> 
> So you do not have impl. done yet.
> 
> I think I told you before, but VS is more permissive than gcc so you see those errors and warnings (warnings are especially useful).

yeah I read it after I wrote this comment. One iplementation is still missing, so I will fix this now, in order to make pushing code to this pull request more readable ;-)

File: main/simulator.cpp [hunk 1]
```diff
@@ -1,9 +1,115 @@
...
+    : frequency_(frequency), numOfRaysPerRow_(numOfRaysPerRow),
+      manager_(SceneManager(collectors, simulationRadius)), depth_(depth),
+      sourcePower_(sourcePower) {
+  // TODO: no implementation yet
+  // model_ = manager_.getModelTriangles();
+
+  if (!manager_.loadModel(objPath)) {
+    std::stringstream ss;
+    ss << "Invalid model path";
+    throw std::invalid_argument(ss.str());
```
Comment (janeusz2000) [inline main/simulator.cpp:15]: When I look at this, I noticed that I could just write std::invalid_argument("invalid model path"). There is no need to use std::stringstream here

File: main/simulator.cpp [hunk 2]
```diff
@@ -1,9 +1,115 @@
...
+        if ((ray.getOrigin() - hitData->origin()).magnitude() <
+            (ray.getOrigin() - closestHitData.origin()).magnitude()) {
+          closestHitData = *hitData;
+          continueRecursion = true;
+        }
+      }
+    }
+
+    // if sphereWall hit point is closer to ray then
+    // Triangle hits or does ray hit Sphere wall only.
```
Comment (janeusz2000) [inline main/simulator.cpp:60]: I need to fix this comment

File: main/simulator.cpp [hunk 3]
```diff
@@ -1,9 +1,115 @@
...
+  }
 }
 
 void Simulator::calculatePressure() {
   // TODO: calculates all pressure from energy collectors
+}
+
+void Simulator::rayTrace(const core::Ray &ray, core::RayHitData *hitData,
+                         int depth) {
+  core::RayHitData closestHitData;
```
Comment (janeusz2000) [inline main/simulator.cpp:43]: This should be inside if statement

File: main/simulator.cpp [hunk 4]
```diff
@@ -1,9 +1,115 @@
...
+  collectors_ = manager_.getEnergyCollectors();
+
+  source_ = generators::PointSource(frequency_, numOfRaysPerRow_,
+                                    manager_.getSmapleSize(), &offsetGenerator_,
+                                    simulationRadius, sourcePower);
+}
+
 void Simulator::run() {
-  // TODO: runs simulation
+  for (size_t xIter = 0; xIter < numOfRaysPerRow_; ++xIter) {
```
Comment (dejw) [inline main/simulator.cpp:28]: Some comments:

1. size_t
2. I am confused about the naming:
  - (small) x**Iter** suggests this is an iterator, while it is not. How about just "x"
  - numRaysPerRow_ is used in both rows and columns.
3. (serious) numOfRaysPerRow_ seems to be **strictly** tight to the PointSource instance you create in constructor, so having it here seems like a violation of the PointSource API. It seems to be like there is only one possible and correct way of generating the rays from the point source (i.e integer from 0 to N, so that everything is within bounds), so why not just moving it to the source itself and calling the new function here? Note that simulator does not care about the x/y coordinates specifically. For example how about doing this:

```c++
Ray ray;
while (point_source.generateRay(&ray)) {
  rayTrace(ray, depth_);
}
```

More general point: I noticed that you very much like writing code so that it does everything at the same time, and rely on side effects to do the work. For example here, `rayTrace` does 2 things: actually bouncing rays and then distributing the energy to collectors, via a side effect (i.e rayTrace does not return anything so we know something has changed, because the members changed).

This style of coding is really hard to follow, and I think it is much better to:
1. follow SRP -- single responsibility principle (read what this is)
2. strive to avoid side-effects, and make sure to return proper values as results of the function calls

In this case it would be much easier to follow if you wrote the code like this;

```c++
Ray ray;
RayHitData sphereHit;
while (point_source.generateRay(&ray)) {
  if (!rayTrace(ray, depth_, &sphereHit)) {
    continue;
  }

  distributeEnergyFromHit(sphereHit);
}
```

Now you can implement distributeEnergyFromHit however you want, and the rayTrace does not have to know how to distribute rays. This is easier to test as well, because currently when calling rayTrace you need to test both bouncing and energy distribution.

In my approach you can test separately, and a bug in one does not affect the other, and it is much easier to create small self-contained test data for each method.

File: main/simulator.cpp [hunk 5]
```diff
@@ -1,9 +1,115 @@
...
+    for (size_t yIter = 0; yIter < numOfRaysPerRow_; ++yIter) {
+      core::Ray currentRay = source_.generateRay(xIter, yIter);
+      core::RayHitData hitData;
```
Comment (dejw) [inline main/simulator.cpp:31]: In this particular case it is probably not an issue, but you are creating a new variable on a stack on each loop iteration. In case of heavier objects, it may cause performance drops (the best is to profile of course, but it is good to have a habit of not doing this).

Move this outside of the loop. Having this outside of the loop, is close enough to where this is first used.

File: main/simulator.h [hunk 6]
```diff
@@ -2,38 +2,50 @@
...
-  Simulator() = delete;
+  // Disallow copy and assing
   Simulator(const Simulator &) = delete;
   Simulator &operator=(const Simulator &) = delete;
 
-  float frequency_;
-  size_t rayNumPerRow_;
+  generators::PointSource source_;
+  // TODO: I want to be able to change engine type when program is running
+  generators::FakeRandomGen offsetGenerator_;
```
Comment (dejw) [inline main/simulator.h:43]: Two words: dependency injection (see my other comment).

File: obj/generators.h [hunk 7]
```diff
@@ -36,12 +44,13 @@ struct FakeRandomGen : public RandomGen {
 // and origin of the source position
 class PointSource final {
 public:
-  PointSource() = delete;
-  explicit PointSource(
-      float freq, int numOfRaysPerRow, float SampleSize, RandomGen *randomGen,
-      core::Vec3 origin = core::Vec3(0, 0, constants::kDefaultSimulationRadius))
+  PointSource() = default;
```
Comment (dejw) [inline obj/generators.h:47]: Why changed from delete to default? Pay up 50 chf!

File: obj/generators.h [hunk 8]
```diff
@@ -36,12 +44,13 @@ struct FakeRandomGen : public RandomGen {
...
+  explicit PointSource(float freq, int numOfRaysPerRow, float SampleSize,
+                       RandomGen *randomGen, float simulationRadius,
+                       float sourcePower)
```
Comment (dejw) [inline obj/generators.h:50]: Why is this called _power_ while this represents an energy. Do you specify the time interval somewhere so that power can be converted to energy? Or did I miss something?

If this is just energy for one ray, why not calling it `rayEnergy`? Also this will increase while your numOfRaysPerRow increase, to please decide if you want to provide total source energy or it is easier to provide per ray value. I am **guessing** it is easier to provide total.

File: obj/objects.h [hunk 9]
```diff
@@ -52,6 +52,7 @@ class Sphere : public Object {
 // point to the sphere collectors;
 class SphereWall : public Sphere {
 public:
+  SphereWall() : SphereWall(constants::kDefaultSimulationRadius / 2){};
```
Comment (dejw) [inline obj/objects.h:55]: +50CHF

File: tests/object_utils_test.cpp [hunk 10]
```diff
@@ -474,12 +24,11 @@ TEST(SphereWallObjectTest, RayInsideSphereHit) {
...
-  ASSERT_TRUE(sphere.hitObject(rayInsideSphere, kSkipFreq, &ignore))
-      << "Ray Hit didn't occur.";
+  Ray rayInsideSphere(sphere.getOrigin(), dir, kDefaultEnergyPerRay);
+  ASSERT_TRUE(sphere.hitObject(rayInsideSphere, kSkipFreq, &ignore));
 
-  Ray rayOutsideSphere(sphere.getOrigin() + Vec3(2 * kR, 0, 0), dir);
-  ASSERT_FALSE(sphere.hitObject(rayOutsideSphere, kSkipFreq, &ignore))
-      << "Ray hit when it shouldn't";
+  Ray rayOutsideSphere(sphere.getOrigin() + Vec3(2 * kR, 0, 0), dir,
+                       kDefaultEnergyPerRay);
```
Comment (dejw) [inline tests/object_utils_test.cpp:31]: Having constants outside, makes this test really hard to understand, why hitObject should return false...

One would have to lookup what kDefaultBlah is and then go back here.

File: tests/simulator_test.cpp [hunk 11]
```diff
@@ -1,2 +1,50 @@
...
+const Vec3 kDefaultSourcePosition(0, 0, constants::kDefaultSimulationRadius);
+const Vec3 kVecStraightDown(0, 0, -1);
+const float kDefaultEnergyPerRay =
+    constants::kDefaultSourcePower /
+    (constants::kDefaultNumRaysPerRow * constants::kDefaultNumRaysPerRow);
+
+class ReferencePlateTest : public ::testing::Test {
+public:
+  ReferencePlateTest()
+      : sim(Simulator(kSkipFreq, kDefaultNumRaysPerRow,
```
Comment (dejw) [inline tests/simulator_test.cpp:26]: Do you realize that you create the temp. object by calling Simulator(...) and then calling copy ctor. by calling: `sim_(Simulator(...))` right? Just drop the `Simulator()` and it will work

Comment (dejw) [inline tests/simulator_test.cpp:26]: BTW: a good rule of thumb is "if you need to pass more than 5 things to a function/ctor, you are doing it wrong (design wise)"

File: tests/simulator_test.cpp [hunk 12]
```diff
@@ -1,2 +1,50 @@
...
+  Simulator sim;
+};
+
+TEST_F(ReferencePlateTest, EnergyCollectionFromRay) {
+
+  RayHitData hitData;
+
+  Ray starightDown(kDefaultSourcePosition, kVecStraightDown,
+                   kDefaultEnergyPerRay);
+  sim.rayTrace(starightDown, &hitData, 2);
```
Comment (dejw) [inline tests/simulator_test.cpp:40]: What 2 means? A good way to documenting call sites is calling out argument names explicitly e.g:

```c++
sim.rayTrace(starightDown, &hitData, /*depth=*/2);
```

clang also warns you in case you get the name wrong, so that there is some additional check that you are passing the right param.

**Review by dejw — CHANGES_REQUESTED** [review]
There are major issues with the code

1. **too tightly coupling between all the classes** --  this is due to both, using constants/globals too much, not using dependency injection; in the current state it is hard to change the real implementation in the test to a fake one that is easier to follow,

2. **the non-trivial code is not really easy to read** -- this is due to relying on side-effects too much (mutating the internals of the class instead of just returning results from functions), and not following the Single Responsibility Principle (SRP); this makes it hard to test things, because everything is interdependent, and you cannot test one single thing.

3. There are also other stylistic issues with the code, that I marked in the comments.

4. You need to pay 100CHF for adding 2x default constructors.

File: BUILD.bazel [hunk 13]
```diff
@@ -2,10 +2,6 @@ load("@rules_cc//cc:defs.bzl", "cc_library", "cc_test")
 
 # https://docs.bazel.build/versions/master/be/c-cpp.html#cc_library
 
-# TODO: create multiple libraries and ad them to test cases. T
```
[the remark this answers is not in the document]
Comment (janeusz2000) [reply BUILD.bazel:5]: This is weird. 
image
I deleted it because I wrote all libraries for every single test

File: constants.h [hunk 14]
```diff
@@ -2,20 +2,29 @@
...
+#include <string>
 
 namespace constants {
 const float kAccuracy = 0.000005;
 const float kSoundSpeed = 343.216; // This is measured in 20'C at 1000 hPa
 const float kPi = std::acos(-1);
-const float kDefaultSimulationRadius = 4; // [m]
-const int kPopulation = 37;
 const float kHitAccuracy = 0.0001;
+const size_t kDefaultNumRaysPerRow = 1000;
```
[the remark this answers is not in the document]
Comment (janeusz2000) [reply constants.h:12]: Im sure, that I won't be having problems with results of the artymetric operation being an invalid sate in this case. However maybe you are right that size_t should be only used in accesing containers such as vector, so maybe it is a good idea to change this to int, just to develop good habits.

File: main/sceneManager.h [hunk 15]
```diff
@@ -16,19 +16,21 @@
...
 
-  bool loadModel(std::string_view objPath);
+  [[nodiscard]] bool loadModel(std::string_view objPath);
 
   std::vector<objects::EnergyCollector *> getEnergyCollectors() const;
   std::vector<objects::TriangleObj *> getModelTriangles() const;
   std::vector<objects::TriangleObj *> getReferencePlate() const;
-  const objects::SphereWall &getSphereWall();
-  const generators::PointSource &getPointSource();
+  objects::SphereWall &getSphereWall() { return sphereWall_; }
```
[the remark this answers is not in the document]
Comment (janeusz2000) [reply main/sceneManager.h:28]: because I'm modifying hitData in hitObject, so this can't be const ;)

File: main/sceneManager.cpp [hunk 16]
```diff
@@ -1,7 +1,8 @@
 #include "main/sceneManager.h"
 #include <sstream>
 SceneManager::SceneManager(int collectors, float simulationRadius)
-    : numCollectors_(collectors), simulationRadius_(simulationRadius) {
+    : numCollectors_(collectors), simulationRadius_(simulationRadius),
+      sphereWall_(objects::SphereWall(simulationRadius / 2)) {
```
[the remark this answers is not in the document]
Comment (janeusz2000) [reply main/sceneManager.cpp:5]: Fixed

File: main/simulator.cpp [hunk 17]
```diff
@@ -1,9 +1,115 @@
 #include "main/simulator.h"
 
+Simulator::Simulator(float frequency, size_t numOfRaysPerRow, int collectors,
+                     float simulationRadius, std::string_view objPath,
+                     int depth, float sourcePower)
+    : frequency_(frequency), numOfRaysPerRow_(numOfRaysPerRow),
+      manager_(SceneManager(collectors, simulationRadius)), depth_(depth),
```
[the remark this answers is not in the document]
Comment (janeusz2000) [reply main/simulator.cpp:7]: fixed c-tor

File: constants.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (dejw) [reply constants.h:12]: In my opinion, you should never use it, even with vectors and such, because this is just asking for trouble.

> "I am sure, that..."

In computer science, there is no such thing. There is a joke that people write "should not happen" comments in the code, and then they happen.

File: main/rayTracer.cpp [hunk 18]
```diff
@@ -0,0 +1,14 @@
+#include "core/ray.h"
+#include "core/vec3.h"
+#include "main/model.h"
+#include "obj/objects.h"
+
+class RayTracer {
```
Comment (dejw) [inline main/rayTracer.cpp:6]: Why is this defined 2x?

File: obj/generators.cpp [hunk 19]
```diff
@@ -1,67 +1,24 @@
...
-  return os << "Point Source: origin: " << pointSource.origin()
-            << ", number of rays per row: " << pointSource.numOfRaysPerRow()
-            << ", diffusor size: " << pointSource.sampleSize()
-            << ", frequency: " << pointSource.frequency();
-}
+PointSpeakerRayFactory::PointSpeakerRayFactory(int numOfRays, Model *model)
+    : numOfRays_(numOfRays), currentRayNum_(0), model_(model) {
 
-// METHODS
+  float simulationHeight = 8 * model->height();
```
Comment (dejw) [inline obj/generators.cpp]: Where those magic formulas and numbers are coming from? Why 8? Why 4 * 7 ?

File: obj/generators.h [hunk 20]
```diff
@@ -4,75 +4,53 @@
...
-  core::Vec3 getLeftCorner() const;
-  void setLeftCorner(const core::Vec3 &point);
+  [[nodiscard]] bool genRay(core::Ray *ray) override;
 
 private:
-  core::Vec3 origin_, leftCorner_;
-  float frequency_, sampleSize_;
-  size_t numOfRaysPerRow_;
+  int numOfRays_, currentRayNum_;
+  Model *model_;
```
Comment (dejw) [inline obj/generators.h]: Do you need to store this here? I thought you only need to keep it in c-tor to get some values out.

File: main/simulator.cpp [hunk 21]
```diff
@@ -1,9 +1,115 @@
...
+                         int depth) {
+  core::RayHitData closestHitData;
+
+  if (depth > 0) {
+    bool continueRecursion = false;
+
+    // Finding the closest object that ray hits
+    for (const auto &obj : model_) {
+      if (obj->hitObject(ray, frequency_, hitData)) {
+        if ((ray.getOrigin() - hitData->origin()).magnitude() <
```
[the remark this answers is not in the document]
Comment (janeusz2000) [reply main/simulator.cpp]: noted.. will be usefull in new implementation ;-)

File: main/simulator.cpp [hunk 22]
```diff
@@ -1,9 +1,115 @@
...
+            (ray.getOrigin() - closestHitData.origin()).magnitude()) {
```
[the remark this answers is not in the document]
Comment (janeusz2000) [reply main/simulator.cpp]: noted ;-)

File: obj/objects.h (hunk 9 above)
Comment (janeusz2000) [reply obj/objects.h]: fixed

File: BUILD.bazel (hunk 13 above)
[the remark this answers is not in the document]
Comment (janeusz2000) [reply BUILD.bazel]: noted that I need to read mroe how to build with bazel

File: obj/generators.h (hunk 20 above)
Comment (janeusz2000) [reply obj/generators.h]: true, fixed

**Comment by janeusz2000** [comment]
> Now it make more sense!
> 
> But it is still lacking. When you will write documentation comment to this function, please explain why it has to be like this, and why they 8xmodelheight etc.

noted 👍
